src/Integration/UserSettingsConfiguration/UserSettingsProvider.cs

```diff
@@ -155,59 +80,40 @@ public UserSettings UserSettings
-    public string GlobalAnalysisSettingsFilePath => globalFilePaths.settingsFile;
-    public string SolutionAnalysisSettingsFilePath => solutionFilePaths?.settingsFile;
-
-    public void DisableRule(string ruleId)
-    {
-        Debug.Assert(!string.IsNullOrEmpty(ruleId), "DisableRule: ruleId should not be null/empty");
-
-        var newRules = UserSettings.AnalysisSettings.Rules.SetItem(ruleId, new RuleConfig(RuleLevel.Off));
-        var globalSettings = new GlobalAnalysisSettings(newRules, UserSettings.AnalysisSettings.GlobalFileExclusions);
-        serializer.SafeSave(GlobalAnalysisSettingsFilePath, globalSettings);
-        SafeClearUserSettingsCache();
-    }
-
-    public void UpdateGlobalFileExclusions(IEnumerable<string> exclusions)
-    {
-        var globalSettings = new GlobalAnalysisSettings(UserSettings.AnalysisSettings.Rules, exclusions.ToImmutableArray());
-        serializer.SafeSave(GlobalAnalysisSettingsFilePath, globalSettings);
-        SafeClearUserSettingsCache();
-    }
 
-    public void UpdateSolutionFileExclusions(IEnumerable<string> exclusions)
-    {
-        var solutionSettings = new SolutionAnalysisSettings(UserSettings.AnalysisSettings.AnalysisProperties, exclusions.ToImmutableArray());
-        serializer.SafeSave(SolutionAnalysisSettingsFilePath, solutionSettings);
-        SafeClearUserSettingsCache();
-    }
-
-    public void UpdateAnalysisProperties(Dictionary<string, string> analysisProperties)
-    {
-        var solutionSettings = new SolutionAnalysisSettings(analysisProperties, UserSettings.AnalysisSettings.SolutionFileExclusions);
-        serializer.SafeSave(SolutionAnalysisSettingsFilePath, solutionSettings);
-        SafeClearUserSettingsCache();
-    }
-
-    public void EnsureGlobalAnalysisSettingsFileExists()
+    public void Dispose()
     {
-        if (!fileSystem.File.Exists(GlobalAnalysisSettingsFilePath))
+        if (!disposed)
         {
-            serializer.SafeSave(GlobalAnalysisSettingsFilePath, new GlobalAnalysisSettings());
+            if (InitializationProcessor.IsFinalized)
+            {
+                solutionSettingsStorage.SettingsFileChanged -= OnSettingsFileChanged;
+                solutionSettingsStorage.Dispose();
+                globalSettingsStorage.SettingsFileChanged -= OnSettingsFileChanged;
+                globalSettingsStorage.Dispose();
+                activeSolutionTracker.ActiveSolutionChanged -= ActiveSolutionTrackerOnActiveSolutionChanged;
+            }
+            disposed = true;
         }
     }
 
-    public void EnsureSolutionAnalysisSettingsFileExists()
+    private void OnSettingsFileChanged(object sender, EventArgs e) => ResetConfiguration();
+
+    private void ActiveSolutionTrackerOnActiveSolutionChanged(object sender, ActiveSolutionChangedEventArgs e)
     {
-        if (SolutionAnalysisSettingsFilePath == null)
+        if (!e.IsSolutionOpen || e.SolutionName == null)
```

Comment:
There is a potential issue here. This event handler needs to be executed strictly after the `SolutionSettingsStorage's` event handler. If not, then when this raises `SettingsChanged` and its event handlers are then requesting to load the new settings, the file path to the solution settings is not initialized. Currently, this is not happening because the event ordering is ensured by this class' dependency on `SolutionSettingsStorage`, however this is would be very easy to miss when modifying these classes. I suggest adding a comment here to explain that.